<a href="https://colab.research.google.com/github/Eshashr/My-Projects/blob/main/DataMiningTechniquesProject1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 1

Project 1 is about using cross-validation to explore hyperparameter options for two different types of models. We are using a random forest model and a k-nearest neighbors model.

We are using another dataset with a long history in machine learning, a dataset from the Univerisity of Wisconsin on breast cancer.

Be sure to mount your drive so that you can access the data set.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_validate
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler

In [ ]:
# Read in the dataset and specify the id column to be row indices
bcwd = pd.read_csv('/content/drive/MyDrive/HDS/Datasets/breast_cancer_wisconsin_diagnostic.csv', index_col='id')
# Recode the classes with numbers
bcwd['diagnosis'].replace('M',1, inplace=True)
bcwd['diagnosis'].replace('B',0, inplace=True)
bcwd.head()

diagnosis  radius_mean  ...  symmetry_worst  fractal_dimension_worst
id                                ...                                         
842302            1        17.99  ...          0.4601                  0.11890
842517            1        20.57  ...          0.2750                  0.08902
84300903          1        19.69  ...          0.3613                  0.08758
84348301          1        11.42  ...          0.6638                  0.17300
84358402          1        20.29  ...          0.2364                  0.07678

[5 rows x 31 columns]

In [ ]:
# Create training and test sets
# Using a seed for the random number generator to ensure the split is the same every time
# for the purposes of this project
training_set, test_set = train_test_split(bcwd, test_size=0.20, random_state=42)

# Use all of the features, excluding diagnosis at the beginning
training_features = training_set.iloc[:,1:]
# Targets are at the end
training_targets = training_set.iloc[:,0]

# Pull out the same features and targets from test set
testing_features = test_set.iloc[:,1:]
testing_targets = test_set.iloc[:,0]

# Use this list of measurements to score your cross-validation results
scoring = ['accuracy','precision', 'recall', 'f1']

# Part One

In this part we will train a series of random forest models using cross-validation. We will perform what is called a grid search of two different hyperparameters. A grid search explores different combinations of values across multiple hyperparamters. 

For our purposes, we will explore max_features and max_depth. Max_features specifies how many features each tree has access too. Max_depth limits the growth of the tree. We will explore two values (small and large) for max_features, 4 and 20, and two values of max_depth, 4 and 10. This means a total of 4 models will be created. 

The first model with max_features=4 and max_depth=5 has code ready to go. You will need to create the other three models that have different pairings of hyperparameter values. Finally, pick your model with the highest test_accuracy and evaluate it on the test set (Code is given for this, but you must set the hyperparameters based on your results).


RFM MODEL 1

In [ ]:

# Sample code of a random forest cross-validation run

# Create the random forest model with max_features=4 and max_depth=4
rf_model = RandomForestClassifier(criterion='entropy', max_features=4, max_depth=4)

# Cross validation calculating multiple metrics
rf_model.fit(training_features, training_targets)
scores = cross_validate(rf_model, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results = pd.DataFrame(scores)
results.mean()

fit_time           0.197134
score_time         0.012426
test_accuracy      0.951648
train_accuracy     0.992857
test_precision     0.957452
train_precision    1.000000
test_recall        0.911230
train_recall       0.980773
test_f1            0.933093
train_f1           0.990290
dtype: float64

RFM MODEL 2

In [ ]:

# Create the random forest model with max_features=4 and max_depth=10
# Sample code of a random forest cross-validation run

rf_model = RandomForestClassifier(criterion='entropy', max_features=4, max_depth=10)

# Cross validation calculating multiple metrics
rf_model.fit(training_features, training_targets)
scores = cross_validate(rf_model, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results = pd.DataFrame(scores)
results.mean()

fit_time           0.201579
score_time         0.011459
test_accuracy      0.958242
train_accuracy     1.000000
test_precision     0.957945
train_precision    1.000000
test_recall        0.928877
train_recall       1.000000
test_f1            0.942316
train_f1           1.000000
dtype: float64

RFM MODEL 3

In [ ]:

# Create the random forest model with max_features=20 and max_depth=4
rf_model = RandomForestClassifier(criterion='entropy', max_features=20, max_depth=4)

# Cross validation calculating multiple metrics
rf_model.fit(training_features, training_targets)
scores = cross_validate(rf_model, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results = pd.DataFrame(scores)
results.mean()

fit_time           0.394423
score_time         0.011787
test_accuracy      0.953846
train_accuracy     0.994505
test_precision     0.951721
train_precision    1.000000
test_recall        0.922995
train_recall       0.985207
test_f1            0.936455
train_f1           0.992543
dtype: float64

RFM MODEL 4

In [ ]:

# Create the random forest model with max_features=20 and max_depth=10
rf_model = RandomForestClassifier(criterion='entropy', max_features=20, max_depth=10)

# Cross validation calculating multiple metrics
rf_model.fit(training_features, training_targets)
scores = cross_validate(rf_model, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results = pd.DataFrame(scores)
results.mean()

fit_time           0.399752
score_time         0.013280
test_accuracy      0.960440
train_accuracy     1.000000
test_precision     0.963675
train_precision    1.000000
test_recall        0.929055
train_recall       1.000000
test_f1            0.945330
train_f1           1.000000
dtype: float64

In [ ]:
# Sample code evaluating an rf model on the test set
# Customize it as appropriate
rf_model = RandomForestClassifier(criterion='entropy', max_features=20, max_depth=10)
rf_model.fit(training_features, training_targets)

# The score method is quick way of getting the accuracy on the testing data
rf_model.score(testing_features, testing_targets)

0.9649122807017544

## Questions on Part One

1) Which model had the highest accuracy test_accuracy score?

Model with max_feature = 20 and Max_depth = 10, RFM Model 4, has the highest test accuracy score of 0.960440.

2) Which model had the highest test_f1 score?

Model with max_feature = 20 and Max_depth = 10, RFM Model 4, has the highest test_f1 score 0.945330.

3) Which model had the highest fit_time? Why do you think this is?

Model with max_feature = 20 and Max_depth = 10, RFM Model 4, has the highest fit_time score 0.399752. I think its because the depth and feature is the highest in this model.

4) Which model had the lowest fit_time? Why do you think this is?

Model with max_feature = 4 and Max_depth = 4, RFM Model 1, has the lowest fit_time score 0.197134. I think its because the depth and feature is the lowest in this model.

5) Does the model with the highest fit_time have the best f1 score?

Yes, RFM Model 4 has the highest fit_time and best f1 score.

6) Did you see any evidence of over-fitting from the training accuracy scores?

In RFM Model 2 and 4, the traning_accuracy is exactly 1.000, so this is an evidence of overfitting.

7) How did your best model in terms of test_accuracy score do when evaluated on the testing set using the score method? How different were the results from the test_accuracy score during cross-validation? 

RFM Model 4, when evaluated through the score method test_accuracy was 0.9649122807017544 compared to 0.960440 using the cross-validation. The results were pretty similar.

# Part Two
In part two we will explore using k-nearest neighbors algorithm. First, we will investigate whether scaling the data improves results. Second, we ill use cross-validation to set the k value of our model. 

Code is given to create and evaluate a k-nearest neighbors model with K=3 on the original data. Code is then given to scale the data. You need to then recreate the knn model with K=3 and evalute its performance on the scaled data. Then create knn models with K=7 and K=12 and evaluate them. Finally, pick your model with the highest test_accuracy and evaluate it on the test set (Code is given for this, but you must set the correct K value).


In [ ]:
# Sample code of a KNN cross-validation run

# Create KNN model with n_neighbors=3, train on original data
knn_model = KNeighborsClassifier(n_neighbors=3)

# Cross validation calculating multiple metrics
knn_model.fit(training_features, training_targets)
scores = cross_validate(knn_model, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results = pd.DataFrame(scores)
results.mean()

fit_time           0.002121
score_time         0.007111
test_accuracy      0.920879
train_accuracy     0.948901
test_precision     0.934580
train_precision    0.964923
test_recall        0.846702
train_recall       0.895000
test_f1            0.886192
train_f1           0.928605
dtype: float64

In [ ]:
# Scale the features to have a mean of 0 and a standard deviation of 1
feature_scaler = StandardScaler().fit(training_features)
training_features_scaled = feature_scaler.transform(training_features)
testing_features_scaled = feature_scaler.transform(testing_features)


KNN Model 1

In [ ]:

# Create KNN model with n_neighbors=3, train on scaled data
knn_model = KNeighborsClassifier(n_neighbors=3)

# Cross validation calculating multiple metrics
knn_model.fit(training_features_scaled, training_targets)
scores = cross_validate(knn_model, training_features_scaled, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results = pd.DataFrame(scores)
results.mean()



fit_time           0.001330
score_time         0.008117
test_accuracy      0.960440
train_accuracy     0.982418
test_precision     0.974531
train_precision    0.996947
test_recall        0.917291
train_recall       0.955621
test_f1            0.944211
train_f1           0.975814
dtype: float64

KNN Model 2

In [ ]:

# Create KNN model with n_neighbors=7, train on scaled data
knn_model = KNeighborsClassifier(n_neighbors=7)

# Cross validation calculating multiple metrics
knn_model.fit(training_features_scaled, training_targets)
scores = cross_validate(knn_model, training_features_scaled, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results = pd.DataFrame(scores)
results.mean()

fit_time           0.001292
score_time         0.008113
test_accuracy      0.960440
train_accuracy     0.971429
test_precision     0.981339
train_precision    0.998374
test_recall        0.911586
train_recall       0.924575
test_f1            0.943823
train_f1           0.959980
dtype: float64

KNN Model 3

In [ ]:

# Create KNN model with n_neighbors=12, train on scaled data
knn_model = KNeighborsClassifier(n_neighbors=12)

# Cross validation calculating multiple metrics
knn_model.fit(training_features_scaled, training_targets)
scores = cross_validate(knn_model, training_features_scaled, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results = pd.DataFrame(scores)
results.mean()

fit_time           0.001357
score_time         0.008502
test_accuracy      0.953846
train_accuracy     0.963187
test_precision     0.993548
train_precision    0.998319
test_recall        0.881818
train_recall       0.902364
test_f1            0.933414
train_f1           0.947875
dtype: float64

In [ ]:
# Sample code evaluating a knn model on the scaled test set
# Customize it as appropriate
knn_model = KNeighborsClassifier(n_neighbors=3)
knn_model.fit(training_features_scaled, training_targets)

# The score method is quick way of getting the accuracy on the testing data
knn_model.score(testing_features_scaled, testing_targets)

0.9473684210526315

## Questions for Part Two

1) Did the results from KNN model change when you switched to scaled features?

Yes, the results changed but not super significantly.

2) Which KNN model had the highest test_accuracy? Did it also have the highest test_f1 score? Is precision or recall higher and what does that say about the model's performance?

KNN Model 1, n_neighbors=3, had the highest test_accuracy of 0.960440. Yes this also has the highest test_f1 score of 0.944211. test_precision is 0.974531
and test_recall is 0.917291, so the precision is higher than the recall.
This means that our model had low number of false positives and returns more relevant results than irrelevant ones (precision), and our model does minimize false negetive buy not as much as precision. In general our system is more accurate if these are both high and it is. 

 It means that our model can measure correctly identifying True Positives than other models (recall high), and ratio between the True Positives and all the Positives is lower than in the other models (precision lower).

3) How did your best model in terms of test_accuracy do when evaluated on the testing set? Did the accuracy go up or down and by how much?

In the training set, the test_accuracy was 0.9473684210526315, lower than 0.960440 by about 0.02.

4) How does your best KNN model compare to your best Random Forest model in terms of accuracy on the testing set? 
On the testing set, my best Random Forest model's accuracy was 0.9649122807017544 and my best KNN's accuracy was 0.9473684210526315. So, KNN's accuracy was lower.

5) How do fit_times compare between the KNN models and the Random Forest models?
Random Forest Models fit_times Range: 0.19 - 0.39 (appoximate)
KNN models fit_times Range: 0.012 - 0.013 (appoximate)
KNN models' fit_times were much lower than that of Random Forest Models.
